### ISSUE TITLE
Feature Request: slow down audio

### ISSUE THREAD

**Issue by nephros** [poetaster/harbour-audiocut#5]
The speed function can currently only speed up clips.

Can you add an option to 'reverse the polarity' of that function?

**Comment by poetaster** [comment]
Yes, it should be possible. The pydub library provides a framerate cv function (ratecv). I'll try that.

**Comment by poetaster** [comment]
Hmmm. I just tested speed-ignore pitch (which of course pitches down) and a setting of 0.5 slows it down, so that works. 

Did you mean that 'speed- keep pitch' doesn't seem to work?

**Comment by nephros** [comment]
> Did you mean that 'speed- keep pitch' doesn't seem to work?

Yes, after checking this again, there seems to be an UI glitch which led me to believe it can't be used to slow down a clip:

If freshly opened, selecting the "keep pitch" mode, the slider is at its lowest value, which is 1.

However, switching to "ignore pitch" allows it to be set lower than 1, and with that being set, one can switch back to "keep pitch" and the value stays below 1.

However from one or two small tests, applying that to a clip does not produce the desired effect - I get garbled audio files if I do.

**Comment by poetaster** [comment]
> > Did you mean that 'speed- keep pitch' doesn't seem to work?
> 
> Yes, after checking this again, there seems to be an UI glitch which led me to believe it can't be used to slow down a clip:
> 
> If freshly opened, selecting the "keep pitch" mode, the slider is at its lowest value, which is 1.
>
> However, switching to "ignore pitch" allows it to be set lower than 1, and with that being set, one can switch back to "keep pitch" and the value stays below 1.
> 

Ok, that's one new ticket for the UI portion ....

> However from one or two small tests, applying that to a clip does not produce the desired effect - I get garbled audio files if I do.

hah! I don't get ANY audio output. It's a div / 0 ! yeah!

```
  File "/usr/share/harbour-audiocut/qml/py/audiox.py", line 257, in speedChange
    middle = extract2.speedup( float(factorSpeed) )
[W] unknown:552 - file:///usr/share/harbour-audiocut/qml/pages/FirstPage.qml:552: RangeError: Invalid Date out of range
[D] unknown:0 - "PyOtherSide error: Traceback (most recent call last):\n\n  File \"/usr/share/harbour-audiocut/qml/py/audiox.py\", line 257, in speedChange\n    middle = extract2.speedup( float(factorSpeed) )\n\n  File \"/usr/share/harbour-audiocut/lib/pydub/effects.py\", line 91, in speedup\n    out = out.append(chunk, crossfade=crossfade)\n\n  File \"/usr/share/harbour-audiocut/lib/pydub/audio_segment.py\", line 1264, in append\n    xf = seg1[-crossfade:].fade(to_gain=-120, start=0, end=float('inf'))\n\n  File \"/usr/share/harbour-audiocut/lib/pydub/audio_segment.py\", line 1357, in fade\n    scale_step = gain_delta / fade_frames\n\nZeroDivisionError: float division by zero\n"
[D] onError:391 - python error: Return value of PyObject call is NULL: Traceback (most recent call last):

  File "/usr/share/harbour-audiocut/qml/py/audiox.py", line 257, in speedChange
    middle = extract2.speedup( float(factorSpeed) )

  File "/usr/share/harbour-audiocut/lib/pydub/effects.py", line 91, in speedup
    out = out.append(chunk, crossfade=crossfade)

  File "/usr/share/harbour-audiocut/lib/pydub/audio_segment.py", line 1264, in append
    xf = seg1[-crossfade:].fade(to_gain=-120, start=0, end=float('inf'))

  File "/usr/share/harbour-audiocut/lib/pydub/audio_segment.py", line 1357, in fade
    scale_step = gain_delta / fade_frames

ZeroDivisionError: float division by zero


  File "/usr/share/harbour-audiocut/lib/pydub/effects.py", line 91, in speedup
    out = out.append(chunk, crossfade=crossfade)

  File "/usr/share/harbour-audiocut/lib/pydub/audio_segment.py", line 1264, in append
    xf = seg1[-crossfade:].fade(to_gain=-120, start=0, end=float('inf'))

  File "/usr/share/harbour-audiocut/lib/pydub/audio_segment.py", line 1357, in fade
    scale_step = gain_delta / fade_frames

ZeroDivisionError: float division by zero

```

**Comment by poetaster** [comment]
So, the div/0 error occurs with selection sizes falling too low and is also actually a UI bug. 

Slowing down without pitch shifting is basically just sampling and splicing. In pydub, by default 150 ms chunks crossfade over 25 ms. The implementation DOES sometimes for some values work. But it can sound 'garbled' for sure. 

I'm trying some other values for chunk and crossfade, (which can get dangerous with largeish files).

**Comment by poetaster** [comment]
Ok, so it's NOT a bug in UI. The pydub library simply can't produce good results with the method it's using. So 'keep the pitch' only works when speeding up. 

There is a library called librosa which is said to be able to do this. And it might be possible with ffmpeg directly. So, I'll take a look at those methods.

**Comment by poetaster** [comment]
I've pushed the feature implemented in b953b6ec1d4a6ba980cf40551dc4a8163cc6e95f
to. https://openrepos.net/content/planetosstore/audiocut

**Comment by poetaster** [comment]
https://github.com/poetaster/harbour-audiocut/releases/tag/1.4.1
